Empty array instead of nil for source_extract

itsWill opened the pull request:
### Summary

The source_extract method will return nil when it can't find the file name in
the backtrace, methods that consume this method expect an array and the nil ends
up causing type errors down the road like it happened here: #36341. This patch
introduces a to_a call that converts the nil into an empty array this makes it so
client classes can correctly handle a empty list of source code lines and not
trigger a type error.

### Other Information

Note this is a bit defensive and preventative imo, this occured for the edge case described here: https://github.com/rails/rails/issues/36341#issuecomment-503673532 and potentially other unknown code paths could still cause this behaviour to occur, but I don't know of any. 

At any rate I do like having the method return a `[]` instead of a nil and thing that is a better interface for this class in general since the method will be returning consistent types. 

discussed this with @rafaelfranca and @gmcgibbon

Commit:
Empty array instead of nil for source_extract

The source_extract method will return nil when it can't find the file name in
the backtrace, methods that consume this method expect an array and the nil ends
up causing type errors down the road like it happened here: #36341. This
patch refactors the source_extract method so that it returns an empty
array instead of nil when it can't find the source code.

Co-authored-by: Kasper Timm Hansen <kaspth@gmail.com>

gmcgibbon commented on actionview/test/template/template_error_test.rb:
```diff
@@ -34,4 +34,20 @@ def test_provides_useful_inspect
   end
+
+
+  class Template
```
Can you define this inline with `Class.new`?

gmcgibbon commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   annotated_source returns an empty array instead of nil when the source code
+    can't be found within a TemplateError
```
Maybe  "Cast result of `annotated_source_code` to array so `TemplateError`s without a template in the backtrace are surfaced properly by `DebugExceptions`." better explains the situation?

gmcgibbon reviewed (commented):
I'm good to merge this, but I want a second opinion as this could be considered overly defensive programming.

kaspth commented on actionview/test/template/template_error_test.rb:
```diff
@@ -34,4 +34,20 @@ def test_provides_useful_inspect
     assert_equal "#<ActionView::Template::Error: original>", error.inspect
   end
+
+
```
✂️ this line

kaspth commented on actionview/test/template/template_error_test.rb:
```diff
@@ -34,4 +34,20 @@ def test_provides_useful_inspect
+
+    error = begin
+      raise Exception.new
+    rescue Exception
```
Would it be clearer with just `raise` and then `rescue` without any specific exception?

itsWill replied:
yes, yes it would!

kaspth commented:
I've been looking into the surrounding code and I don't `source_extract` is right. It uses `formatted_code_for` which could return a Hash if `output` was `:html`. However it never seems to be anything but the default `:console` from `source_extract` (because `source_extract` is only called from `annotated_source_code`.

My take is that we should simplify `formatted_code_for` to this while we're here:

```ruby
def formatted_code_for(source_code, line_counter, indent)
  indent_template = "%#{indent}s: %s"

  source_code.map.with_index(line_counter) do |line, counter|
    indent_template % [counter, line]
  end
end
```

Otherwise calling `to_a` on the Hash would also break the code. And if not, the Hash would this when `concat(annotated_source_code)` is hit:

```irb
>> [].concat({})
Traceback (most recent call last):
        5: from /Users/kaspth/.rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        4: from /Users/kaspth/.rbenv/versions/2.6.3/bin/irb:23:in `load'
        3: from /Users/kaspth/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):2
        1: from (irb):2:in `concat'
TypeError (no implicit conversion of Hash into Array)
```

Finally `source_extract` should have it's `output` argument removed. In fact, maybe we should just inline `source_extract` into `annotated_source_code`. I think we need to dig deeper here than just going for `to_a` and calling it a day — I'm very happy you're looking into this! 😊

itsWill commented:
@kaspth good catch :+1: 

I've removed the `output` parameter and cleaned up the code some, so we return `[]` instead of `nil` so we no longer need the `to_a`. 

This is a much cleaner solution ❤️

gmcgibbon commented on actionview/test/template/template_error_test.rb:
```diff
@@ -34,4 +34,19 @@ def test_provides_useful_inspect
+    rescue
+      raise ActionView::Template::Error.new(template) rescue $!
+    end
+    assert_equal [], error.annotated_source_code
```
Put a newline above ☝️

gmcgibbon commented on actionview/lib/action_view/template/error.rb:
```diff
@@ -122,15 +122,11 @@ def source_location
-          start_value = (output == :html) ? {} : []
-          source_code.inject(start_value) do |result, line|
+        def formatted_code_for(source_code, line_counter, indent)
+          indent_template = "%#{indent}s: %s"
```
Is this an `indent_template` or a `line_template`?

itsWill replied:
The line of the template is being indented by a `indent` amount, so I went with indent template as a better name for what this code is doing, but I'm happy to change it to `line_template` if you feel that it's better.

gmcgibbon commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   annotated_source_code returns an empty array so TemplateErrors without a
+    template in the backtrace are surfaced properly by DebugExceptions.
+
+    *Guilherme Mansur*
```
If you are giving credit to @kaspth in the commit, his name should go here as well. 😄

itsWill replied:
good catch! fixed.

kaspth commented:
❤️💪
